File: leetcode/solutions/0404.SumOfLeftLeaves/sumOfLeftLeaves.go
package solution

// TreeNode indicates a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func sumOfLeftLeaves(root *TreeNode) int {
	return dfs(root, false)
}

func dfs(node *TreeNode, isLeft bool) int {
	if node == nil {
		return 0
	}

	left, right := node.Left, node.Right

	if left == nil && right == nil && isLeft {
		return node.Val
	}

	return dfs(left, true) + dfs(right, false)
}

File: leetcode/solutions/0404.SumOfLeftLeaves/sumOfLeftLeaves.ts
import { BinaryTreeNode } from "../../../datastructure/BinaryTreeNode.ts";

export default function sumOfLeftLeaves(
  root: BinaryTreeNode<number> | null,
  isLeft: boolean = false
): number {
  if (!root) {
    return 0;
  }

  const { left, right } = root;

  if (isLeft && !left && !right) {
    return root.val;
  }

  return sumOfLeftLeaves(left, true) + sumOfLeftLeaves(right, false);
}
